src/plugins/layout/shadowdom.ts

```diff
@@ -26,39 +25,33 @@ export class ShadowDom {
     return node as IShadowDomNode;
   }
 
-  public insertShadowNode(node: Node, parentIndex: number, nextSiblingIndex: number, layout?: ILayoutState): IShadowDomNode {
+  public insertShadowNode(node: Node, parentIndex: number, nextSiblingIndex: number): IShadowDomNode {
     let isDocument = (node === document);
     let index = this.setNodeIndex(node);
     let parent = (isDocument ? this.shadowDomRoot : this.getShadowNode(parentIndex)) as IShadowDomNode;
     let nextSibling = this.getShadowNode(nextSiblingIndex);
     let shadowNode = this.doc.createElement("div") as IShadowDomNode;
-    let ignore = (node === document) ? false : parent && parent.ignore;
     shadowNode.id = "" + index;
     shadowNode.node = node;
-    shadowNode.layout = layout;
-    shadowNode.ignore = (layout && layout.tag === IgnoreTag) || ignore;
+    shadowNode.ignore = shouldIgnoreNode(node, this);
 
     if (isDocument) {
       this.shadowDocument = shadowNode;
     }
 
-    assert(!!parent, "insertShadowNode", "parent is missing");
-    if (!parent) {
-      return;
-    }
-
     if (this.classifyNodes) {
       this.setClass(shadowNode, NewNodeClassName);
-      if (!this.hasClass(parent, NewNodeClassName)) {
-        this.setClass(shadowNode, TopNewNodeClassName);
-      }
     }
 
-    if (nextSibling) {
-      parent.insertBefore(shadowNode, nextSibling);
-    } else {
-      parent.appendChild(shadowNode);
+    assert(!!parent, "insertShadowNode", "parent is missing");
```

Comment:
Yes, when we insert a node, we always attached it to some already existing node. Even if the parent was removed from DOM in some previous mutation, its shadow node would still be in the ShadowDOM's special 'removed' container, until we process all the mutations, so it should be accessible until the very end.

---
In reply to: [131237635](https://github.com/Microsoft/clarity-js/pull/25#discussion_r131237635) [](ancestors = 131237635)